### PR TITLE
[SPMD] Upstream iter_move_grads_and_optimizers

### PR DESCRIPTION
**Pull request by fegin**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #98785

This PR upstreams `iter_move_grads_and_optimizer` which delay some of the gradients and the corresponding optimizer to the next iteration. D44512863(credit to @lessw2020 ) is the internal implementation, which is only good for the old _SPMD expansion.  This PR changes the implmentation to use the new APIs.

Differential Revision: [D44836486](https://our.internmc.facebook.com/intern/diff/D44836486/)

### CODE REVIEW

File: test/distributed/_spmd/test_transformation.py
Comment (mrshenli) [inline test/distributed/_spmd/test_transformation.py:263]: same comment regarding accuracy tests as previous PRs

Comment (fegin) [reply test/distributed/_spmd/test_transformation.py]: Done

File: torch/distributed/_spmd/graph_optimization.py
Comment (mrshenli) [inline torch/distributed/_spmd/graph_optimization.py:911]: this is very specific to adam. Let's add a TODO/FIXME to track

File: torch/distributed/_spmd/graph_utils.py
Comment (mrshenli) [inline torch/distributed/_spmd/graph_utils.py:116]: nit: simplification?

```python
nodes = reversed(graph.nodes) if reverse_order else graph.nodes
```

**Review by mrshenli — APPROVED** [review]
This is not user-facing yet, stamp to unblock. Please fix Lint before land

File: torch/distributed/_spmd/graph_optimization.py
Comment (fegin) [reply torch/distributed/_spmd/graph_optimization.py:900]: You are right about the usage of this pass. This enables the ability to move nodes around but the algorithm is not decided yet. Right now the API is intended to be used for manual tuning.

File: torch/distributed/_spmd/graph_utils.py
Comment (fegin) [reply torch/distributed/_spmd/graph_utils.py:116]: This is less easier to read after typing hints. Will keep the original form.

**Pull request closed without merging**
